### PR TITLE
Feature: return comments with the parser

### PR DESCRIPTION
**Pull request by Janther**
added the `comments` option and when it's set to `true`, the AST will have an extra property with an array of all of the comments within the source.
Also fixed a bug where the `loc` of a token will display the wrong `end` property if the token contained new lines.

Once this is published `prettier-plugin-solidity` won't need `solidity-comments-extractor` anymore.

### CODE REVIEW

File: src/tokens.ts [hunk 1]
```diff
@@ -40,27 +41,58 @@ function getTokenType(value: string) {
   }
 }
 
+function range(token: AntlrToken): [number, number] {
+  return [token.start, token.stop + 1]
+}
+
+function loc(token: AntlrToken): Location {
+  const textInLines = token.text?.split(/\r?\n/) ?? ['']
```
Comment (fvictorio) [inline src/tokens.ts:49]: Personal preference, feel free to ignore.

```suggestion
  const tokenText = token.text ?? ''
  const textInLines = tokenText.split(/\r?\n/)
```

**Review by fvictorio — APPROVED** [review]
This is great, thanks a lot @Janther!

I left two very optional comments, feel free to merge whether you apply them or not.

**Comment by tonisives** [comment]
Just saying the comment parsing seems to be incorrect for some functions. For instance, this one:
```solidity
    /**
        @notice Returns square-root spot price, lower and upper bounds of the AMM position. 
     */
    function getAMMState()
```
[link](https://github.com/sherlock-audit/2024-03-arrakis/blob/main/valantis-hot/src/HOT.sol#L349)

**Comment by fvictorio** [comment]
@tonisives what does the parser produce and why do you think it's wrong? I tried it and got:

```
[
  {
    type: 'BlockComment',
    value: '*\n' +
      '        @notice Returns square-root spot price, lower and upper bounds of the AMM position. \n' +
      '     '
  }
]
```

which looks correct to me.

**Comment by tonisives** [comment]
I am not familiar with AST representation of comments. I can only write my preference.     

```solidity
*
            @notice Returns square-root spot price, lower and upper bounds of the AMM position. 
         
```

If I look at this result, then it is missing * and / \ characters.


Another issue is with /// function comments.
They are missing 2 of the / in front of the comment

screenshot-2024-06-06-at-16 33 31
[source](https://github.com/sherlock-audit/2024-03-arrakis/blob/main/arrakis-modular/src/ArrakisMetaVaultFactory.sol#L121)

I would also like to get the whole comment (all 9 lines), but they are as separate objects in the result.

I think it is normal to expect that /// lines above the function compose the full function comment. But again, I don't know AST comments format at all.

**Pull request merged**